File: Test_suite/C/graph.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <time.h>

#define NUM_NODES 1000
#define MAX_WEIGHT 10
#define EDGE_PROBABILITY 0.5

typedef struct {
    int node;
    int weight;
} Edge;

typedef struct {
    Edge edges[NUM_NODES];
    int numEdges;
} AdjacencyList;

typedef struct {
    AdjacencyList adjList[NUM_NODES];
} Graph;

void addEdge(Graph *graph, int from, int to, int weight) {
    int fromIndex = graph->adjList[from].numEdges++;
    graph->adjList[from].edges[fromIndex] = (Edge){to, weight};

    int toIndex = graph->adjList[to].numEdges++;
    graph->adjList[to].edges[toIndex] = (Edge){from, weight};
}

void generateLargeGraph(Graph *graph, int numNodes, double edgeProbability, int maxWeight) {
    srand(time(NULL));
    for (int i = 0; i < numNodes; i++) {
        for (int j = i + 1; j < numNodes; j++) {
            if ((double)rand() / RAND_MAX < edgeProbability) {
                int weight = rand() % maxWeight + 1;
                addEdge(graph, i, j, weight);
            }
        }
    }
}

// Simple priority queue (min heap) implementation
typedef struct {
    int node;
    int distance;
} QueueNode;

typedef struct {
    QueueNode nodes[NUM_NODES];
    int size;
} PriorityQueue;

void swap(QueueNode *a, QueueNode *b) {
    QueueNode temp = *a;
    *a = *b;
    *b = temp;
}

void heapify(PriorityQueue *pq, int index) {
    int smallest = index;
    int left = 2 * index + 1;
    int right = 2 * index + 2;

    if (left < pq->size && pq->nodes[left].distance < pq->nodes[smallest].distance)
        smallest = left;
    if (right < pq->size && pq->nodes[right].distance < pq->nodes[smallest].distance)
        smallest = right;

    if (smallest != index) {
        swap(&pq->nodes[index], &pq->nodes[smallest]);
        heapify(pq, smallest);
    }
}

void push(PriorityQueue *pq, QueueNode node) {
    pq->nodes[pq->size] = node;
    int i = pq->size;
    pq->size++;

    while (i != 0 && pq->nodes[(i - 1) / 2].distance > pq->nodes[i].distance) {
        swap(&pq->nodes[(i - 1) / 2], &pq->nodes[i]);
        i = (i - 1) / 2;
    }
}

QueueNode pop(PriorityQueue *pq) {
    if (pq->size <= 0) {
        return (QueueNode){-1, -1};
    }

    QueueNode root = pq->nodes[0];
    pq->nodes[0] = pq->nodes[--pq->size];
    heapify(pq, 0);

    return root;
}

// Dijkstra's algorithm
void dijkstra(Graph *graph, int start, int distances[NUM_NODES]) {
    for (int i = 0; i < NUM_NODES; i++) {
        distances[i] = INT_MAX;
    }
    distances[start] = 0;

    PriorityQueue pq = {0};
    push(&pq, (QueueNode){start, 0});

    while (pq.size > 0) {
        QueueNode minNode = pop(&pq);
        int u = minNode.node;

        for (int i = 0; i < graph->adjList[u].numEdges; i++) {
            Edge edge = graph->adjList[u].edges[i];
            int v = edge.node;
            int weight = edge.weight;

            if (distances[u] + weight < distances[v]) {
                distances[v] = distances[u] + weight;
                push(&pq, (QueueNode){v, distances[v]});
            }
        }
    }
}

int main() {
    Graph graph = {0};
    generateLargeGraph(&graph, NUM_NODES, EDGE_PROBABILITY, MAX_WEIGHT);

    int distances[NUM_NODES];

    clock_t start_time = clock();
    dijkstra(&graph, 0, distances);
    clock_t end_time = clock();

    double execution_time = (double)(end_time - start_time) / CLOCKS_PER_SEC;

    // Estimate memory usage
    size_t graphMemory = NUM_NODES * (NUM_NODES - 1) * sizeof(Edge);
    size_t pqMemory = NUM_NODES * sizeof(QueueNode);
    size_t totalMemory = graphMemory + pqMemory; // Add more if needed for other variables

    printf("Execution Time: %f seconds\n", execution_time);
    printf("Estimated Memory Usage: %zu M\n", totalMemory/(1024*1024));

    return 0;
}

// https://www.onlinegdb.com/online_c_compiler

File: Test_suite/Rust/graph.rs
use rand::rngs::ThreadRng;
use rand::{Rng, thread_rng};
use std::cmp::Ordering;
use std::collections::BinaryHeap;
use std::mem;

const NUM_NODES: usize = 1000;
const MAX_WEIGHT: i32 = 10;
const EDGE_PROBABILITY: f64 = 0.5;

#[derive(Copy, Clone, Eq, PartialEq)]
struct Edge {
    node: usize,
    weight: i32,
}

struct AdjacencyList {
    edges: Vec<Edge>,
}

struct Graph {
    adj_list: Vec<AdjacencyList>,
}

impl Graph {
    fn new() -> Graph {
        let mut adj_list = Vec::with_capacity(NUM_NODES);
        for _ in 0..NUM_NODES {
            adj_list.push(AdjacencyList { edges: Vec::new() });
        }
        Graph { adj_list }
    }

    fn add_edge(&mut self, from: usize, to: usize, weight: i32) {
        self.adj_list[from].edges.push(Edge { node: to, weight });
        self.adj_list[to].edges.push(Edge { node: from, weight });
    }
}

fn generate_large_graph(rng: &mut ThreadRng) -> Graph {
    let mut graph = Graph::new();
    for i in 0..NUM_NODES {
        for j in i + 1..NUM_NODES {
            if rng.gen::<f64>() < EDGE_PROBABILITY {
                let weight = rng.gen_range(1..=MAX_WEIGHT);
                graph.add_edge(i, j, weight);
            }
        }
    }
    graph
}



#[derive(Copy, Clone, Eq, PartialEq)]
struct State {
    cost: i32,
    position: usize,
}

impl Ord for State {
    fn cmp(&self, other: &State) -> Ordering {
        other.cost.cmp(&self.cost).reverse()
    }
}

impl PartialOrd for State {
    fn partial_cmp(&self, other: &State) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn dijkstra(graph: &Graph, start: usize) -> Vec<i32> {
    let mut dist: Vec<_> = (0..NUM_NODES).map(|_| i32::MAX).collect();
    let mut heap = BinaryHeap::new();

    dist[start] = 0;
    heap.push(State { cost: 0, position: start });

    while let Some(State { cost, position }) = heap.pop() {
        if cost > dist[position] { continue; }

        for edge in &graph.adj_list[position].edges {
            let next = State { cost: cost + edge.weight, position: edge.node };

            if next.cost < dist[next.position] {
                heap.push(next);
                dist[next.position] = next.cost;
            }
        }
    }
    dist
}

fn estimate_memory_usage(graph: &Graph) -> usize {
    let mut total_size = 0;

    // Size of the adjacency list vector itself
    total_size += mem::size_of::<Vec<AdjacencyList>>();

    for adj_list in &graph.adj_list {
        // Size of each adjacency list
        total_size += mem::size_of::<AdjacencyList>();
        // Size of the edges vector in each adjacency list
        total_size += mem::size_of::<Vec<Edge>>();
        // Size of the edges in each adjacency list
        total_size += adj_list.edges.capacity() * mem::size_of::<Edge>();
    }

    total_size
}

fn main() {
    let mut rng = thread_rng();
    let graph = generate_large_graph(&mut rng);

    let start_time = std::time::Instant::now();
    let distances = dijkstra(&graph, 0);
    let end_time = std::time::Instant::now();

    println!("Execution Time: {:?}", end_time.duration_since(start_time));

    // Example usage of distances
    // for (i, &dist) in distances.iter().enumerate() {
    //     println!("Distance from node 0 to node {}: {}", i, dist);
    // }
    let memory_usage = estimate_memory_usage(&graph);
    println!("Estimated Memory Usage: {} M", memory_usage/(1024*1024));
}
//https://play.rust-lang.org/?version=stable&mode=debug&edition=2021
